packages/@aws-cdk/aws-ecs/lib/drain-hook/instance-drain-hook.ts

```diff
@@ -0,0 +1,101 @@
+import autoscaling = require('@aws-cdk/aws-autoscaling');
+import iam = require('@aws-cdk/aws-iam');
+import lambda = require('@aws-cdk/aws-lambda');
+import sns = require('@aws-cdk/aws-sns');
+import cdk = require('@aws-cdk/cdk');
+import path = require('path');
+import { ICluster } from '../cluster';
+
+// Reference for the source in this package:
+//
+// https://github.com/aws-samples/ecs-refarch-cloudformation/blob/master/infrastructure/lifecyclehook.yaml
+
+/**
+ * Properties for instance draining hook
+ */
+export interface InstanceDrainHookProps {
+  /**
+   * The AutoScalingGroup to install the instance draining hook for
+   */
+  autoScalingGroup: autoscaling.IAutoScalingGroup;
+
+  /**
+   * The cluster on which tasks have been scheduled
+   */
+  cluster: ICluster;
+
+  /**
+   * How many seconds to give tasks to drain before the instance is terminated anyway
+   *
+   * Must be between 0 and 900.
+   *
+   * @default 900
+   */
+  drainTimeSeconds?: number;
+}
+
+/**
+ * A hook to drain instances from ECS traffic before they're terminated
+ */
+export class InstanceDrainHook extends cdk.Construct {
+  constructor(parent: cdk.Construct, id: string, props: InstanceDrainHookProps) {
+    super(parent, id);
+
+    const drainTimeSeconds = props.drainTimeSeconds !== undefined ? props.drainTimeSeconds : 300;
+
+    if (drainTimeSeconds < 0 || drainTimeSeconds > 900) {
+      throw new Error(`Drain time must be between 0 and 900 seconds, got: ${drainTimeSeconds}`);
+    }
+
+    // Invoke Lambda via SNS Topic
+    const topic = new sns.Topic(this, 'Topic');
+    const fn = new lambda.Function(this, 'Function', {
+      code: lambda.Code.directory(path.join(__dirname, 'lambda-source')),
```

Comment:
- .directory is deprecated (use `.asset`)
- Also, for L2s, if possible, I think we should prefer inline code instead of assets to reduce the surface area

Reply:
I totally agree!